File: prisma/seed.ts
import "dotenv/config";
import bcrypt from "bcryptjs";
import { db } from "../src/lib/db";

const DEV_EMAIL = "demo@example.com";
const DEV_PASSWORD = "password123";

async function main() {
  const passwordHash = await bcrypt.hash(DEV_PASSWORD, 10);

  const user = await db.user.upsert({
    where: { email: DEV_EMAIL },
    update: {},
    create: { email: DEV_EMAIL, name: "Demo Rep", passwordHash },
  });

  const workspace = await db.workspace.upsert({
    where: { id: "demo-workspace" },
    update: {},
    create: {
      id: "demo-workspace",
      name: "Acme Sales",
      defaultCurrency: "USD",
    },
  });

  await db.workspaceMember.upsert({
    where: { workspaceId_userId: { workspaceId: workspace.id, userId: user.id } },
    update: {},
    create: { workspaceId: workspace.id, userId: user.id, role: "owner" },
  });

  const pipeline = await db.pipeline.upsert({
    where: { id: "demo-pipeline" },
    update: {},
    create: {
      id: "demo-pipeline",
      workspaceId: workspace.id,
      name: "New Business",
      isDefault: true,
    },
  });

  const stageDefs = [
    { id: "demo-stage-prospecting", name: "Prospecting", order: 0, probability: 10 },
    { id: "demo-stage-qualified", name: "Qualified", order: 1, probability: 30 },
    { id: "demo-stage-proposal", name: "Proposal", order: 2, probability: 60 },
    { id: "demo-stage-negotiation", name: "Negotiation", order: 3, probability: 80 },
    { id: "demo-stage-won", name: "Closed Won", order: 4, probability: 100, isWon: true },
    { id: "demo-stage-lost", name: "Closed Lost", order: 5, probability: 0, isLost: true },
  ];

  for (const stage of stageDefs) {
    await db.stage.upsert({
      where: { id: stage.id },
      update: {},
      create: { ...stage, pipelineId: pipeline.id },
    });
  }

  const company = await db.company.upsert({
    where: { id: "demo-company" },
    update: {},
    create: {
      id: "demo-company",
      workspaceId: workspace.id,
      name: "Globex Corporation",
      domain: "globex.example",
    },
  });

  const contact = await db.contact.upsert({
    where: { id: "demo-contact" },
    update: {},
    create: {
      id: "demo-contact",
      workspaceId: workspace.id,
      companyId: company.id,
      firstName: "Jane",
      lastName: "Doe",
      email: "jane.doe@globex.example",
      lifecycleStage: "opportunity",
    },
  });

  await db.deal.upsert({
    where: { id: "demo-deal" },
    update: {},
    create: {
      id: "demo-deal",
      workspaceId: workspace.id,
      pipelineId: pipeline.id,
      stageId: "demo-stage-qualified",
      companyId: company.id,
      amountCents: 4_500_000,
      currency: "USD",
      contacts: { connect: [{ id: contact.id }] },
    },
  });

  const activityDefs = [
    {
      id: "demo-activity-1",
      type: "call",
      body: "Intro call with Jane. Interested in the Enterprise tier, wants a security review before signing.",
    },
    {
      id: "demo-activity-2",
      type: "email",
      body: "Sent pricing breakdown and a comparison against their current vendor.",
    },
  ];
  for (const activity of activityDefs) {
    await db.activity.upsert({
      where: { id: activity.id },
      update: {},
      create: {
        id: activity.id,
        workspaceId: workspace.id,
        dealId: "demo-deal",
        companyId: company.id,
        type: activity.type,
        body: activity.body,
      },
    });
  }

  console.log(`Seeded. Sign in with ${DEV_EMAIL} / ${DEV_PASSWORD}`);
}

main()
  .catch((err) => {
    console.error(err);
    process.exitCode = 1;
  })
  .finally(async () => {
    await db.$disconnect();
  });

File: backend/CrmApi/Data/Seed.cs
using CrmApi.Models;
using Microsoft.EntityFrameworkCore;

namespace CrmApi.Data;

public static class Seed
{
    private const string DevEmail = "demo@example.com";
    private const string DevPassword = "password123";

    public static async Task RunAsync(AppDbContext db)
    {
        var user = await db.Users.FirstOrDefaultAsync(u => u.Email == DevEmail);
        if (user is null)
        {
            user = new User
            {
                Email = DevEmail,
                Name = "Demo Rep",
                PasswordHash = BCrypt.Net.BCrypt.HashPassword(DevPassword),
            };
            db.Users.Add(user);
        }

        var workspace = await db.Workspaces.FirstOrDefaultAsync(w => w.Name == "Acme Sales");
        if (workspace is null)
        {
            workspace = new Workspace { Name = "Acme Sales" };
            db.Workspaces.Add(workspace);
        }
        await db.SaveChangesAsync();

        if (!await db.WorkspaceMembers.AnyAsync(m => m.WorkspaceId == workspace.Id && m.UserId == user.Id))
        {
            db.WorkspaceMembers.Add(new WorkspaceMember { WorkspaceId = workspace.Id, UserId = user.Id, Role = "owner" });
        }

        var pipeline = await db.Pipelines.FirstOrDefaultAsync(p => p.WorkspaceId == workspace.Id && p.IsDefault);
        if (pipeline is null)
        {
            pipeline = new Pipeline { WorkspaceId = workspace.Id, Name = "New Business", IsDefault = true };
            db.Pipelines.Add(pipeline);
            await db.SaveChangesAsync();

            var stageDefs = new (string Name, int Order, int Probability, bool IsWon, bool IsLost)[]
            {
                ("Prospecting", 0, 10, false, false),
                ("Qualified", 1, 30, false, false),
                ("Proposal", 2, 60, false, false),
                ("Negotiation", 3, 80, false, false),
                ("Closed Won", 4, 100, true, false),
                ("Closed Lost", 5, 0, false, true),
            };
            foreach (var s in stageDefs)
            {
                db.Stages.Add(new Stage
                {
                    PipelineId = pipeline.Id,
                    Name = s.Name,
                    Order = s.Order,
                    Probability = s.Probability,
                    IsWon = s.IsWon,
                    IsLost = s.IsLost,
                });
            }
            await db.SaveChangesAsync();
        }

        var company = await db.Companies.FirstOrDefaultAsync(c => c.WorkspaceId == workspace.Id && c.Name == "Globex Corporation");
        if (company is null)
        {
            company = new Company { WorkspaceId = workspace.Id, Name = "Globex Corporation", Domain = "globex.example" };
            db.Companies.Add(company);
            await db.SaveChangesAsync();
        }

        var contact = await db.Contacts.FirstOrDefaultAsync(c => c.WorkspaceId == workspace.Id && c.Email == "jane.doe@globex.example");
        if (contact is null)
        {
            contact = new Contact
            {
                WorkspaceId = workspace.Id,
                CompanyId = company.Id,
                FirstName = "Jane",
                LastName = "Doe",
                Email = "jane.doe@globex.example",
                LifecycleStage = "opportunity",
            };
            db.Contacts.Add(contact);
            await db.SaveChangesAsync();
        }

        var qualifiedStage = await db.Stages.FirstAsync(s => s.PipelineId == pipeline.Id && s.Name == "Qualified");
        var deal = await db.Deals.FirstOrDefaultAsync(d => d.WorkspaceId == workspace.Id && d.CompanyId == company.Id);
        if (deal is null)
        {
            deal = new Deal
            {
                WorkspaceId = workspace.Id,
                PipelineId = pipeline.Id,
                StageId = qualifiedStage.Id,
                CompanyId = company.Id,
                AmountCents = 4_500_000,
                Currency = "USD",
                Contacts = [contact],
            };
            db.Deals.Add(deal);
            await db.SaveChangesAsync();
        }

        if (!await db.Activities.AnyAsync(a => a.DealId == deal.Id))
        {
            db.Activities.AddRange(
                new Activity
                {
                    WorkspaceId = workspace.Id,
                    DealId = deal.Id,
                    CompanyId = company.Id,
                    Type = "call",
                    Body = "Intro call with Jane. Interested in the Enterprise tier, wants a security review before signing.",
                },
                new Activity
                {
                    WorkspaceId = workspace.Id,
                    DealId = deal.Id,
                    CompanyId = company.Id,
                    Type = "email",
                    Body = "Sent pricing breakdown and a comparison against their current vendor.",
                }
            );
            await db.SaveChangesAsync();
        }

        Console.WriteLine($"Seeded. Sign in with {DevEmail} / {DevPassword}");
    }
}
